### PR TITLE
Feat: Added inputs variables dereferencing support [CFG-1501]

### PR DESCRIPTION
**Pull request by YairZ101**
#### Adds supporting reading variable values in Terraform (.tf) files.

This is done to help on our vision to create better TF support for our product and find more issues that might have been ignored so far because they would be coming from variable values.
This PR is to support input variables, with or without default values. 

### Background context

In the [Decision Document](https://www.notion.so/snyk/Terraform-Variable-Dereferencing-in-the-Snyk-CLI-24c2c15f5a0942429b7454b0796a4298), we decided to create a new library in Go where the implementation for this would live.

We already have played [a spike](https://www.notion.so/snyk/POC-Golang-based-Terraform-var-dereferencing-and-Terraform-to-JSON-conversion-with-merging-all-cont-cc931b9f965347ab826f0b7076b676cf#9e93f530bb1f4d74bd85055a706b948e) on how this could look like.

We plan to build a parser that can deal with (end goal):
- Variable blocks inside HCL (.tf) files
- Local blocks inside HCL (.tf) files
- Variables with default values
- Variables in the terraform.tfvars default file
- A definition of input variables can be found [here](https://www.terraform.io/language/values/variables).

**As part of this ticket**, we will only be adding support for input variables (and defaults).

### Additional information

- [Jira Issue CFG-1501](https://www.terraform.io/language/values/variables)
-  [Slack thread](https://snyk.slack.com/archives/C030WL9BLSW/p1644402193378949)

### CODE REVIEW

File: terraform/constants.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package terraform
+
+const (
+	TF           = ".tf"
+	TF_VARS      = ".tfvars"
+	TF_VARS_JSON = ".tfvars.json"
```
Comment (ipapast) [inline terraform/constants.go:6]:
```suggestion
	TF           = ".tf"
```
nit: we are not really using these yet, do we?

**Review by ipapast — APPROVED** [review]
I left two small comments, non blocking. I think it looks good in general and it works! I tested this commit with the `test.js` from Snyk CLI locally as well and the integration works and still returning results for the defaults. 🙌

File: terraform/hcl2json_test.go [hunk 2]
```diff
@@ -531,48 +532,98 @@ block "label_one" {
 }
 
 func TestExtractVariablesSuccess(t *testing.T) {
```
Comment (ipapast) [inline terraform/hcl2json_test.go:534]: maybe a missing test here, we could have a test case with a duplicate variable in the fileContent. we should fail and get invalid hcl I believe

Comment (ipapast) [reply terraform/hcl2json_test.go:534]: Actually this is more of an error case of the hcl2json case and not the extract itself. we can also add them in a seperate PR

File: terraform/diagnostics.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package terraform
+
+import (
+	"log"
+
+	"github.com/hashicorp/hcl/v2"
+)
+
+func handleDiagnostics(issue string, diags hcl.Diagnostics, filename string) {
```
Comment (teodora-sandu) [inline terraform/diagnostics.go:9]: This functionality will be used in the CLI, so I expect what will happen is wherever we use this function in this code it will log directly to the user's output. That is too much logging in my opinion and it will clutter the CLI output a lot

In the CLI we have debug logs for this sort of thing so we don't want to log directly. I think having debug logs would be nice but we need to figure this out a bit more before doing it. Do you think we should do this now or can we create a ticket so add debug logs and so think of a better solution for this?

Comment (YairZ101) [reply terraform/diagnostics.go:9]: good point we didn't thought about that.
I think that creating a separate ticket to tackle debug logs sounds like a great idea.

File: terraform/variables.go [hunk 4]
```diff
@@ -1,25 +1,64 @@
...
+func extractFromTfFile(filename string, file *hcl.File) (VariableMap, hcl.Diagnostics) {
+	varMap := VariableMap{}
+	bodyContent, hclDiags := file.Body.Content(tfFileSchema)
+
+	handleDiagnostics("Validation issue", hclDiags, filename)
+	if hclDiags.HasErrors() {
+		return VariableMap{}, hclDiags
+	}
+
+	for _, block := range bodyContent.Blocks {
```
Comment (teodora-sandu) [inline terraform/variables.go:42]: Is this going to run on all types of blocks? A file might have more than just `variable` blocks, so for example if I add the following test:
```
{
			name: "Non-variable block",
			input: TestInput{
				fileName: "test.tf",
				fileContent: `
provider "google" {
  project = "acme-app"
  default  = "us-central1"
}`,
			},
			expected: VariableMap{
				"var": cty.ObjectVal(VariableMap{
					"google": cty.StringVal("us-central1"),
				}),
				"local": cty.ObjectVal(VariableMap{
					"dummy": cty.StringVal("dummy"),
				}),
			},
		},
```

It passes although `google` is not a variable

File: terraform/variables.go [hunk 5]
```diff
@@ -1,25 +1,64 @@
...
-	// implement "var" when we do input variables and default values
+type VariableMap map[string]cty.Value
+
+func extractFromFile(filename string, file *hcl.File) (VariableMap, hcl.Diagnostics) {
+	// TODO: remove temporary dummy locals
 	// implement "local" when we do local values
+	varMap := VariableMap{}
+	var hclDiags hcl.Diagnostics
+
+	if strings.HasSuffix(filename, TF) {
```
[the remark this answers is not in the document]
Comment (YairZ101) [reply terraform/variables.go:18]: >if there was another reason you were thinking about doing the check here, happy to hear, I might be also missing something I haven't thought of :D

The reason we did this check here is because `.tfvars` files will have a different logic to extract vars from them.
If this condition wasn't here than the logic will fail when we will try to extract vars from `.tfvars` files.

File: terraform/constants.go (hunk 1 above)
Comment (YairZ101) [reply terraform/constants.go:6]: we don't, those are leftovers from when we looked at extracting vars from `.tfvars` files.

File: terraform/variables.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (YairZ101) [reply terraform/variables.go:18]: please correct me if I'm wrong (I joined in the middle so I might missed some stuff), but we are sending to the `ParseModule` function also `.tfvars` files no? AFAIK they are part of a TF module.
If we do than we need this condition to "filter" out the `.tfvars` files.

File: terraform/diagnostics.go (hunk 3 above)
Comment (ipapast) [reply terraform/diagnostics.go:9]: Ah nice! @teodora-sandu sounds like you have some thoughts already, can you open a ticket about this?

File: terraform/hcl2json_test.go (hunk 2 above)
Comment (ipapast) [reply terraform/hcl2json_test.go:534]: terraform validate will fail. We don't need a test for this in the TestExtractVariablesFailure cases, sorry for the confusion. This was the point of my second comment, we could add one in the failure cases of hcl2json instead, up to us on how much we want to unit test edge cases.

File: terraform/variables.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (YairZ101) [reply terraform/variables.go:18]: okay I understand, in this case it does make sense to remove this condition.

**Comment by YairZ101** [comment]
@ipapast and @teodora-sandu can you please do a quick re-review to the changes I did following your feedback? :)

File: terraform/variables.go [hunk 6]
```diff
@@ -5,21 +5,51 @@ import (
...
 		"local": cty.ObjectVal(VariableMap{
-			"dummy": cty.StringVal("dummy_local"),
+			"dummy": cty.StringVal("dummy"),
 		}),
-	}, nil
+	}, hclDiags
 }
 
-type VariableMap map[string]cty.Value
+func extractFromTfFile(file *hcl.File) (VariableMap, hcl.Diagnostics) {
```
Comment (teodora-sandu) [inline terraform/variables.go:27]: nit: Would it be nicer to name this `extractVariableBlocks` or something like this?

File: terraform/hcl2json_test.go [hunk 7]
```diff
@@ -531,48 +532,115 @@ block "label_one" {
...
+	type TestInput struct {
+		fileName    string
```
Comment (teodora-sandu) [inline terraform/hcl2json_test.go:536]: nit: Do we need the `fileName` here? It was defaulting to `"test"` before and the simpler this test table is the better

**Review by teodora-sandu — APPROVED** [review]
LGTM 🎉  Left some minor nits but up to you if you want to implement them. Great job!

File: terraform/hcl2json_test.go (hunk 7 above)
Comment (YairZ101) [reply terraform/hcl2json_test.go:536]: We added it to prepare for the next ticket where we will have different logic for `.tf` and `.tfvars` files.

File: terraform/variables.go (hunk 6 above)
Comment (YairZ101) [reply terraform/variables.go:27]: I think that `extractVariableBlocks` sounds more abstract but I personally think that `extractFromTfFile` makes it more explicit what this function does and what it's being used for.
